out/notebooks/spl2/installer.ts

```diff
@@ -0,0 +1,514 @@
+        if (termsStatus === TermsAcceptanceStatus.DeclinedForever) {
+            reject(
+                `User opted out of SPL2. To reset this adjust the '${configKeyAcceptedTerms}' ` +
+                `setting to = '${TermsAcceptanceStatus.DeclinedOnce}' in the Splunk Extension Settings.`
+            );
+            return;
+        }
+        // Check for compatible Java version installed already
+        let javaLoc;
+        try {
+            javaLoc = workspace.getConfiguration().get(configKeyJavaPath);
+        } catch (err) {
+            reject(`Error retrieving configuration '${configKeyJavaPath}', err: ${err}`);
+        }
+        // If java hasn't been set up, check $JAVA_HOME before downloading a JDK
+        if (!javaLoc && process.env.JAVA_HOME) {
+            let javaHomeBin = path.join(process.env.JAVA_HOME, 'bin', 'java');
+            if (process.platform === 'win32') {
+                javaHomeBin = `${javaHomeBin}.exe`;
+            }
+            if (isJavaVersionCompatible(javaHomeBin)) {
+                javaLoc = javaHomeBin;
+                try {
+                    workspace.getConfiguration().update(configKeyJavaPath, javaHomeBin, true);
+                } catch (err) {
+                    reject(`Error updating configuration '${configKeyJavaPath}', err: ${err}`);
+                }
+            }
+        }
+
+        // Check workspace for current installed LSP version
+        let lspVersion;
+        try {
+            lspVersion = workspace.getConfiguration().get(configKeyLspVersion);
+        } catch (err) {
+            reject(`Error retrieving configuration '${configKeyLspVersion}', err: ${err}`);
+        }
+        if (javaLoc && lspVersion) {
+            // Already set up, no need to continue
+            resolve(false);
+        }
+        // Setup local storage directory for downloads and installs
+        try {
+            makeLocalStorage(context);
+        } catch (err) {
+            reject(`Error creating local artifact storage for SPL2, err: ${err}`);
+        }
+        
+        let installedLatestLsp = false;
+        if (!lspVersion) {
+            // If we haven't set up a Language Server version prompt use to accept terms
+            // and also confirm install of java if needed
+            try {
+                const accepted = await promptToDownloadLsp(!javaLoc);
+                if (!accepted) {
+                    return;
+                }
+                // Remove any existing LSP artifacts first
+                const localLspDir = getLocalLspDir(context);
+                fs.rmdirSync(localLspDir, { recursive: true });
+                makeLocalStorage(context); // recreate directory
+            
+                await getLatestSpl2Release(context, progressBar);
+                installedLatestLsp = true;
+            } catch (err) {
+                reject(`Error retrieving latest SPL2 release, err: ${err}`);
+            }
+        } else if (!javaLoc) {
```

Comment:
are these two conditions `lspLoc` and `javaLoc` mutually exclusive? What if the user has installed lsp (outdated version) but down the line they uninstalled java?

Reply:
I think I have all cases covered here - in the `promptToDownloadLsp` function we ask the user to also accept downloading the JDK if it's missing. If only the JDK is missing (not lsp) then we ask them to approve downloading the JSK in `promptToDownloadJava`.
|                  |  has lsp |needs lsp|
|-----------|----------|---------|
|  has jdk  |    ✔️    |  1st `if`  |
|needs jdk  |2nd `if`  |    1st `if` |

Reply:
the reason for the slightly convoluted overlap here is to not force the user to accept multiple dialogs for java and lsp to save some steps for them